---
description: "Consecutive non-null assertion operators are unnecessary."
title: "emptyMappingKeys"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="yaml" rule="emptyMappingKeys" />

Empty keys are invalid in YAML and may cause parsers to reject the document or misinterpret its structure.
Even if allowed by a parser, empty keys can be confusing for developers and lead to accidental mistakes in code.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```yaml
: "value"
```

</TabItem>
<TabItem label="✅ Correct">

```yaml
"key": "value"
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## Equivalents in Other Linters

<RuleEquivalents pluginId="yaml" ruleId="emptyMappingKeys" />
