---
description: "Prefer using classList.toggle() over conditional classList.add() and classList.remove()."
title: "classListToggles"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";

import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="browser" rule="classListToggles" />

The `classList.toggle()` method provides a more concise and expressive way to conditionally add or remove a CSS class.
Using conditional logic with `classList.add()` and `classList.remove()` creates unnecessary code duplication and makes the intent less clear.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```js
if (condition) {
	element.classList.add("active");
} else {
	element.classList.remove("active");
}

if (isVisible) {
	button.classList.remove("hidden");
} else {
	button.classList.add("hidden");
}
```

</TabItem>
<TabItem label="✅ Correct">

```js
element.classList.toggle("active", condition);

button.classList.toggle("hidden", !isVisible);
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you need to support very old browsers that do not implement the two-parameter version of `classList.toggle()` (Internet Explorer 11 and earlier), you might need to use the conditional `add()`/`remove()` pattern.
For modern browsers and applications, prefer `toggle()`.

## Further Reading

- [MDN: `Element.classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList)
- [MDN: `DOMTokenList.toggle()`](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle)

## Equivalents in Other Linters

<RuleEquivalents pluginId="browser" ruleId="classListToggles" />
