---
description: "Reports elements that require alt text but are missing it."
title: "altTexts"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="altTexts" />

Alternative text provides a textual description for images and other media.
Search engine crawlers and screen readers use this text to describe the element to users who cannot see it.

Elements like `<img>`, `<area>`, `<input type="image">`, and `<object>` must have appropriate alternative text through `alt`, `aria-label`, `aria-labelledby`, or `title` attributes.

This is required for WCAG 1.1.1 compliance.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<img src="foo.jpg" />
```

```tsx
<img src="foo.jpg" alt />
```

```tsx
<area href="#" />
```

```tsx
<input type="image" src="submit.png" />
```

```tsx
<object data="movie.mp4" />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<img src="foo.jpg" alt="A foo" />
```

```tsx
<img src="foo.jpg" alt="" />
```

```tsx
<img src="foo.jpg" aria-label="Foo" />
```

```tsx
<area alt="Click here" href="#" />
```

```tsx
<input type="image" alt="Submit" />
```

```tsx
<object aria-label="Video" />
```

</TabItem>
</Tabs>

## When Not To Use It

If you're not working with images or media elements, or if you use a framework that automatically injects known good alt text for those elements, you can disable this rule.

## Further Reading

- [WCAG 1.1.1: Non-text Content](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html)
- [WebAIM: Alternative Text](https://webaim.org/techniques/alttext)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="altTexts" />
