---
description: "Reports anchor elements without accessible content."
title: "anchorContent"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="anchorContent" />

Automated tools such as screen readers and search engine crawlers rely on anchors (links) having contents to understand those links.
If anchor is missing accessible content, then it will not be properly understood.

This rule checks that all anchors either have child content or an ARIA property describing the anchor.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<a />
```

```tsx
<a></a>
```

```tsx
<a>
	<span aria-hidden />
</a>
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<a>Link text</a>
```

```tsx
<a aria-label="Link" />
```

```tsx
<a title="Link title" />
```

</TabItem>
</Tabs>

## When Not To Use It

If you are using a framework that automatically generates accessible content for anchors, you can safely disable this rule.

## Further Reading

- [WCAG 2.4.4 - Link Purpose](https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context)
- [WCAG 4.1.2 - Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="anchorContent" />
