---
description: "Reports invalid usage of anchor elements."
title: "anchorValidity"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="anchorValidity" />

Anchor elements should be used for navigation with valid `href` attributes. For interactive elements without navigation, use button elements instead.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<a />
```

```tsx
<a href="#" />
```

```tsx
<a href="javascript:void(0)" />
```

```tsx
<a onClick={() => {}} />
```

```tsx
<a href="#" onClick={() => {}} />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<a href="https://example.com" />
```

```tsx
<a href="/path" />
```

```tsx
<a href="#section" />
```

```tsx
<button onClick={() => {}} />
```

```tsx
<a href="https://example.com" onClick={() => {}} />
```

</TabItem>
</Tabs>

## When Not To Use It

If you have specific requirements for anchor usage that differ from standard web accessibility practices, you may need to disable this rule.

## Further Reading

- [WCAG 2.4.4 - Link Purpose](https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context)
- [WCAG 4.1.2 - Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="anchorValidity" />
