---
description: "Reports anchor elements without accessible content."
title: "anchorContent"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="anchorContent" />

Anchor elements must have content that is accessible to screen readers. Empty anchors or anchors with only hidden content cannot be properly announced.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<a />
```

```tsx
<a></a>
```

```tsx
<a>
	<span aria-hidden />
</a>
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<a>Link text</a>
```

```tsx
<a aria-label="Link" />
```

```tsx
<a title="Link title" />
```

```tsx
<a dangerouslySetInnerHTML={{ __html: "foo" }} />
```

</TabItem>
</Tabs>

## When Not To Use It

If you are not concerned with accessibility, you can disable this rule.

## Further Reading

- [WCAG 2.4.4 - Link Purpose](https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context)
- [WCAG 4.1.2 - Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="anchorContent" />
