---
description: "Reports elements with aria-hidden='true' that are focusable."
title: "ariaHiddenFocusables"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="ariaHiddenFocusables" />

Elements with `aria-hidden="true"` should not be reachable via keyboard navigation.
When users can focus elements they cannot perceive with a screen reader, it creates confusion and unexpected behavior.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<div aria-hidden="true" tabIndex="0" />
```

```tsx
<input aria-hidden="true" />
```

```tsx
<a href="/" aria-hidden="true" />
```

```tsx
<button aria-hidden="true" />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<div aria-hidden="true" />
```

```tsx
<img aria-hidden="true" />
```

```tsx
<button aria-hidden="true" tabIndex="-1" />
```

```tsx
<a href="/" />
```

</TabItem>
</Tabs>

## When Not To Use It

If you have specific use cases where aria-hidden elements need to be focusable (though this is rare and generally not recommended), you can disable this rule.

## Further Reading

- [aria-hidden elements do not contain focusable elements](https://dequeuniversity.com/rules/axe/html/4.4/aria-hidden-focus)
- [Element with aria-hidden has no content in sequential focus navigation](https://www.w3.org/WAI/standards-guidelines/act/rules/6cfa84/proposed/)
- [MDN: aria-hidden](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="ariaHiddenFocusables" />
