---
description: "Reports invalid ARIA roles."
title: "ariaRoles"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="ariaRoles" />

ARIA roles must be valid according to the WAI-ARIA specification.
Invalid role values cannot be used, as they have no meaning to assistive technologies and can cause accessibility issues.
This rule reports on any DOM element given a JSX `role` attribute with an invalid value.

Required for WCAG 4.1.2 compliance.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<div role="datepicker" />
```

```tsx
<div role="range" />
```

```tsx
<div role="" />
```

```tsx
<div role="custom-role" />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<div role="button" />
```

```tsx
<div role="navigation" />
```

```tsx
<div role={dynamicRole} />
```

```tsx
<div />
```

```tsx
<div role="dialog" />
```

```tsx
<div role="none" />
```

</TabItem>
</Tabs>

## When Not To Use It

If you're using custom ARIA roles or working with experimental features, you may need to disable this rule.

## Further Reading

- [WCAG 4.1.2: Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)
- [WAI-ARIA Role Definitions](https://www.w3.org/TR/wai-aria/#role_definitions)
- [MDN: ARIA Roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="ariaRoles" />
