---
description: "Reports invalid or abstract ARIA roles."
title: "ariaRoleValidity"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="ariaRoleValidity" />

ARIA roles must be valid according to the WAI-ARIA specification.
Abstract roles and invalid role values cannot be used, as they have no meaning to assistive technologies.
This rule reports on any DOM element given a JSX `role` attribute with an invalid value.

This is required for WCAG 4.1.2 compliance.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<div role="datepicker" />
```

```tsx
<div role="range" />
```

```tsx
<div role="" />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<div role="button" />
```

```tsx
<div role="navigation" />
```

```tsx
<div role={dynamicRole} />
```

```tsx
<div />
```

</TabItem>
</Tabs>

## When Not To Use It

If you're using custom ARIA roles or working with experimental features, you may need to disable this rule.

## Further Reading

- [WCAG 4.1.2: Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)
- [WAI-ARIA Role Definitions](https://www.w3.org/TR/wai-aria/#role_definitions)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="ariaRoleValidity" />
