---
description: "Reports autoFocus props that are not set to false."
title: "autoFocusProps"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="autoFocusProps" />

Auto-focusing elements can cause usability issues for both sighted and non-sighted users.
It can be disruptive to users who rely on screen readers or keyboard navigation, as it moves focus unexpectedly.

Consider letting users focus elements manually instead.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<div autoFocus />
```

```tsx
<div autoFocus="true" />
```

```tsx
<input autoFocus={true} />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<div />
```

```tsx
<div autoFocus="false" />
```

```tsx
<div autoFocus={false} />
```

</TabItem>
</Tabs>

## When Not To Use It

If you have specific use cases where auto-focus is necessary for user experience (such as modal dialogs), you may want to disable this rule for those specific cases.

## Further Reading

- [WHATWG HTML Standard: The autofocus attribute](https://html.spec.whatwg.org/multipage/interaction.html#attr-fe-autofocus)
- [The accessibility of HTML 5 autofocus](https://www.brucelawson.co.uk/2009/the-accessibility-of-html-5-autofocus/)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="autoFocusProps" />
