---
description: "Reports mouse events without corresponding keyboard events."
title: "mouseEventKeyEvents"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="mouseEventKeyEvents" />

Mouse events should have keyboard equivalents for users who cannot use a mouse. This ensures that functionality is accessible to users with physical disabilities, AT compatibility, and screen reader users.

`onMouseOver` must be paired with `onFocus`, and `onMouseOut` must be paired with `onBlur`.

This is required for WCAG 2.1.1 compliance.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<div onMouseOver={() => void 0} />
```

```tsx
<div onMouseOut={() => void 0} />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<div onMouseOver={() => void 0} onFocus={() => void 0} />
```

```tsx
<div onMouseOut={() => void 0} onBlur={() => void 0} />
```

</TabItem>
</Tabs>

## When Not To Use It

If your application doesn't use mouse event handlers, you can disable this rule.

## Further Reading

- [WCAG 2.1.1: Keyboard](https://www.w3.org/WAI/WCAG21/Understanding/keyboard)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="mouseEventKeyEvents" />
