---
description: "Reports ARIA roles that have semantic HTML element equivalents."
title: "roleTags"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="roleTags" />

Semantic HTML elements have built-in accessibility features and better browser support than generic elements with ARIA roles. Use native HTML elements when available instead of generic elements with role attributes.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<div role="button" />
```

```tsx
<div role="img" />
```

```tsx
<span role="link" />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<button />
```

```tsx
<img />
```

```tsx
<a href="#" />
```

```tsx
<div role="presentation" />
```

</TabItem>
</Tabs>

## When Not To Use It

If you have specific styling or behavioral requirements that prevent using semantic elements, you may need to disable this rule.

## Further Reading

- [WAI-ARIA Roles Model](https://www.w3.org/TR/wai-aria-1.0/roles)
- [MDN: WAI-ARIA Roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="roleTags" />
