---
description: "Reports scope props on non-th elements."
title: "scopeProps"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="jsx" rule="scopeProps" />

The `scope` attribute defines whether a table header is a column header or row header.
Using it on non-`<th>` elements has no semantic meaning and may confuse assistive technologies.

This is required for WCAG 1.3.1 and 4.1.1 compliance.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```tsx
<div scope />
```

```tsx
<div scope="col" />
```

```tsx
<td scope="row" />
```

</TabItem>
<TabItem label="✅ Correct">

```tsx
<th scope="col" />
```

```tsx
<th scope="row" />
```

```tsx
<th scope={scope} />
```

```tsx
<div />
```

</TabItem>
</Tabs>

## When Not To Use It

If you're not working with HTML tables, you can disable this rule.

## Further Reading

- [WCAG 1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships)
- [WCAG 4.1.1: Parsing](https://www.w3.org/WAI/WCAG21/Understanding/parsing)
- [axe-core: scope-attr-valid](https://dequeuniversity.com/rules/axe/3.5/scope-attr-valid)

## Equivalents in Other Linters

<RuleEquivalents pluginId="jsx" ruleId="scopeProps" />
