---
description: "Reports JSON values that are unsafe for data interchange."
title: "valueSafety"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="json" rule="valueSafety" />

JSON is widely used for data interchange between systems, but certain values can cause interoperability issues when transferred between different parsers and environments.
This rule detects values that are unsafe for interchange, preventing data corruption, silent failures, or inconsistent behavior across different platforms and languages.

## Unsafe Values

This rule detects the following unsafe values:

- **Lone surrogates in strings**: Incomplete Unicode character pairs that can cause encoding/decoding failures
- **Numbers that evaluate to Infinity**: Values like `1e400` that exceed JavaScript's number range
- **Unintentional zeros**: Very small numbers (e.g., `1e-400`) that silently evaluate to zero due to precision limitations
- **Unsafe integers**: Numbers outside JavaScript's safe integer range (`Â±2^53-1`) that lose precision
- **Subnormal numbers**: Very small floating point values that may be handled differently across systems

## Examples

<Tabs>
<TabItem label="âŒ Incorrect">

```json
[2e308]
```

```json
{
	"value": "\ud83d"
}
```

```json
[1e-400]
```

```json
{
	"id": 9007199254740992
}
```

```json
[2.2250738585072009e-308]
```

</TabItem>
<TabItem label="âœ… Correct">

```json
[123, 1234, 12345]
```

```json
{
	"emoji": "ðŸ”¥"
}
```

```json
{
	"emoji": "\ud83d\udd25"
}
```

```json
[0, 0.0, 0]
```

```json
{
	"id": 9007199254740991
}
```

</TabItem>
</Tabs>

## When Not To Use It

If your project uses control flow structures that create confusing side effects, such as legacy systems that rely on these potentially problematic values for specific functionality, you might not be able to enable this rule.
Consider refactoring to use safer value representations or thoroughly testing handling of these edge cases in your controlled environment.

## Further Reading

- [UTF-16 Surrogates (Wikipedia)](https://en.wikipedia.org/wiki/UTF-16)
- [Subnormal Numbers (Wikipedia)](https://en.wikipedia.org/wiki/Subnormal_number)
- [Number.MAX_SAFE_INTEGER - JavaScript | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)
- [Number.MIN_SAFE_INTEGER - JavaScript | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MIN_SAFE_INTEGER)

## Equivalents in Other Linters

<RuleEquivalents pluginId="json" ruleId="valueSafety" />
