---
description: "Prefer block-style mappings over flow-style mappings."
title: "blockMappings"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="yml" rule="blockMappings" />

Flow-style mappings (e.g., `{key: value}`) can be harder to read and maintain, especially with nested structures.
Block-style mappings use indentation to show structure, making the YAML more readable and consistent with common YAML conventions.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```yml
key: { nested: value }
```

```yml
outer: { inner: value, another: data }
```

</TabItem>
<TabItem label="✅ Correct">

```yml
key:
  nested: value
```

```yml
outer:
  inner: value
  another: data
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## Further Reading

- [YAML Specification: Flow Mappings](https://yaml.org/spec/1.2.2/#flow-mappings)
- [YAML Specification: Block Mappings](https://yaml.org/spec/1.2.2/#block-mappings)

## Equivalents in Other Linters

<RuleEquivalents pluginId="yml" ruleId="blockMappings" />
