---
description: "Reports empty mapping keys."
title: "emptyMappingKeys"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="yml" rule="emptyMappingKeys" />

Empty keys are invalid in YAML and may cause parsers to reject the document or misinterpret its structure.
Even if allowed by a parser, empty keys can be confusing for developers and lead to accidental mistakes in code.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```yml
: "value"
```

</TabItem>
<TabItem label="✅ Correct">

```yml
"key": "value"
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## Further Reading

- [YAML Specification: Mapping](https://yaml.org/spec/1.2.2/#21-collections)
- [YAML Best Practices](https://yamllint.readthedocs.io/en/stable/)

## Equivalents in Other Linters

<RuleEquivalents pluginId="yml" ruleId="emptyMappingKeys" />
