---
description: "Reports empty documents in YAML files."
title: "emptyDocuments"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="yml" rule="emptyDocuments" />

Empty documents in YAML contain no content and can be confusing or indicate incomplete structure.
While technically valid in some contexts, empty documents often result from mistakes or incomplete editing.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```yml
---
```

```yml
---
```

```yml
---
---
key: value
```

</TabItem>
<TabItem label="✅ Correct">

```yml
key: value
```

```yml
---
key: value
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## Further Reading

- [YAML Specification - Document Markers](https://yaml.org/spec/1.2.2/#22-structures)

## Equivalents in Other Linters

<RuleEquivalents pluginId="yml" ruleId="emptyDocuments" />
