---
description: "Consecutive non-null assertion operators are unnecessary."
title: "emptyMappingKeys"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import RuleEquivalentsTable from "~/components/RuleEquivalentsTable.astro";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="yml" rule="emptyMappingKeys" />

Empty keys are invalid in YAML and may cause parsers to reject the document or misinterpret its structure.
Even if allowed by a parser, empty keys can be confusing for developers and lead to accidental mistakes in code.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```yml
: "value"
```

</TabItem>
<TabItem label="✅ Correct">

```yml
"key": "value"
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## Equivalents in Other Linters

<RuleEquivalentsTable pluginId="yml" ruleId="emptyMappingKeys" />
