---
description: "Class declarations should not be reassigned."
title: "classAssignments"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="classAssignments" />

Reassigning a class declaration can lead to unexpected behavior and makes code harder to understand.
Class declarations are typically intended to be immutable references, and reassigning them can cause confusion about which class is being used.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
class MyClass {}
MyClass = 0;
```

```ts
class Counter {}
Counter++;
```

```ts
class Data {}
Data = null;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
class MyClass {}
const instance = new MyClass();
```

```ts
class Counter {}
const value = Counter;
```

```ts
class MyClass {}
const derived = class extends MyClass {};
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your project uses class reassignments as a design pattern, you might not be able to immediately enable this rule.
Consider using differently-named variables instead to gain mutable references to classes.

## Further Reading

- [MDN documentation on classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)
- [MDN JavaScript Reference: Assignment operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators#assignment_operators)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="classAssignments" />
