---
description: "Reports class methods that don't use `this` and could be static."
title: "classMethodsThis"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="classMethodsThis" />

Class methods that don't use `this` can often be made static.
Static methods make it clear that the method doesn't depend on instance state and can be called without creating an instance of the class.

This rule excludes:

- Constructor methods
- Getter and setter accessors
- Abstract methods
- Methods with the `override` modifier

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
class Calculator {
	add(a: number, b: number) {
		return a + b;
	}

	greet() {
		console.log("hello");
	}
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
class Calculator {
	static add(a: number, b: number) {
		return a + b;
	}

	static greet() {
		console.log("hello");
	}
}

class Example {
	value = 0;

	getValue() {
		return this.value;
	}
}
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you prefer to keep methods as instance methods for consistency or API design reasons, you may want to disable this rule.
Some frameworks or libraries may also require instance methods for proper binding or serialization.

## Further Reading

- [MDN: Static methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static)
- [TypeScript: Classes](https://www.typescriptlang.org/docs/handbook/2/classes.html)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="classMethodsThis" />
