---
description: "Require a flint-enable comment for every flint-disable comment."
title: "directivePairs"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="directivePairs" />

`flint-disable` directive comments disable Flint rules in all lines after the comment.
If you forget to add a corresponding `flint-enable` directive comment, you may overlook legitimate linting issues.

This rule warns when a `flint-disable` directive does not have a matching `flint-enable` directive.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
// flint-disable someRule
const value = 1;
```

```ts
// flint-disable
const value = 1;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
// flint-disable someRule
const value = 1;
// flint-enable someRule
```

```ts
// flint-disable
const value = 1;
// flint-enable
```

```ts
// flint-disable-next-line someRule
const value = 1;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you intentionally disable rules for entire files without enabling them again, you may disable this rule.
Consider using more targeted disable comments like `flint-disable-next-line` instead.

## Further Reading

- [Flint directive comments documentation](https://flint.fyi/docs)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="directivePairs" />
