---
description: "This parameter overrides the previous parameter of the same name."
title: "duplicateArguments"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="duplicateArguments" />

In JavaScript, duplicate parameter names in function signatures can lead to unexpected behavior.
In strict mode (which TypeScript uses by default), duplicate parameter names are a syntax error.
Even in non-strict mode, only the last parameter with a given name is accessible, making earlier parameters with the same name unreachable.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
function calculateTotal(value, value) {
	return value * 2;
}
```

```ts
function process(first, second, first) {
	return first + second;
}
```

```ts
const handler = (data, data) => {
	console.log(data);
};
```

</TabItem>
<TabItem label="✅ Correct">

```ts
function calculateTotal(value) {
	return value * 2;
}
```

```ts
function process(first, second, third) {
	return first + second + third;
}
```

```ts
const handler = (data) => {
	console.log(data);
};
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your codebase relies on non-standard behavior that involves duplicate argument names (which is highly discouraged), you might choose to disable this rule.
For example, if you target a legacy runtime with non-standard JavaScript semantics, standard practices may not apply to you.

## Further Reading

- [MDN: Functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions)
- [MDN: Strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="duplicateArguments" />
