---
description: "Read-only global variables should not be reassigned or modified."
title: "globalAssignments"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="globalAssignments" />

Global variables like `undefined`, `NaN`, `Infinity`, and built-in objects like `Object` and `Array` are read-only and should not be modified.
Attempting to reassign these globals can lead to confusing behavior and potential bugs in your code.
In strict mode, reassigning these globals will throw a TypeError at runtime.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
undefined = 1;

NaN = 42;

Infinity = 100;

Object = null;

Array = function () {};
```

```ts
let count = 5;
undefined += count;

NaN++;

--Infinity;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
let myUndefined = 1;

const notANumber = 42;

const infiniteValue = 100;

let myObject = null;

const myArray = function () {};
```

```ts
let count = 5;
let undefinedValue = undefined;
undefinedValue = count;

let notANumber = NaN;
notANumber++;

let infiniteValue = Infinity;
--infiniteValue;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your codebase intentionally assigns to global objects and is guaranteed to run in an environment that allows that, you might not want to enable this rule.
For example, if you target a legacy runtime with legacy non-standard JavaScript semantics, modern standard practices may not apply to you.

## Further Reading

- [MDN documentation on global objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects)
- [MDN JavaScript Reference: Strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="globalAssignments" />
