---
description: "For-in loops should be guarded with a filtering condition to avoid iterating over inherited properties."
title: "guardedForIns"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="guardedForIns" />

A for-in loop iterates over all enumerable properties of an object, including those inherited from its prototype chain.
Without a filtering condition (such as `hasOwnProperty`), the loop may process properties you did not intend to include.
This can lead to unexpected behavior when the object's prototype has been modified or when dealing with objects from external sources.

Consider adding a filtering condition inside the loop body, such as `if (Object.hasOwn(obj, key))` or `if (obj.hasOwnProperty(key))`.
Alternatively, use `Object.keys(obj)`, `Object.entries(obj)`, or `Object.values(obj)` if you only need the object's own properties.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const data = { a: 1, b: 2 };

for (const key in data) {
	console.log(key, data[key]);
}
```

```ts
function process(object: Record<string, unknown>) {
	for (const property in object) {
		doSomething(property);
	}
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const data = { a: 1, b: 2 };

for (const key in data) {
	if (Object.hasOwn(data, key)) {
		console.log(key, data[key]);
	}
}
```

```ts
function process(object: Record<string, unknown>) {
	for (const property in object) {
		if (object.hasOwnProperty(property)) {
			doSomething(property);
		}
	}
}
```

```ts
const data = { a: 1, b: 2 };

for (const key of Object.keys(data)) {
	console.log(key, data[key]);
}
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you are certain that the objects you iterate over will never have inherited properties, or if your code intentionally relies on iterating over inherited properties, you can turn off this rule.
You might consider using [Flint disable comments](/) and/or [configuration file disables](/) for those specific situations instead of completely disabling this rule.

## Further Reading

- [MDN: for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in)
- [MDN: Object.hasOwn()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn)
- [MDN: Object.prototype.hasOwnProperty()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="guardedForIns" />
