---
description: "Use object literal notation {} or Object.create instead of calling Object as a constructor."
title: "objectCalls"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="objectCalls" />

Calling `Object` as a constructor with `new Object()` is unnecessarily verbose and less idiomatic than using object literal syntax.
Object literal notation `{}` is the preferred and more concise way to create plain objects in JavaScript and TypeScript.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const emptyObject = new Object();
```

```ts
const config = new Object({ key: "value" });
```

```ts
function createEmpty() {
	return new Object();
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const emptyObject = {};
```

```ts
const config = { key: "value" };
```

```ts
function createEmpty() {
	return {};
}
```

```ts
const objectWithoutPrototype = Object.create(null);
```

```ts
const copy = Object.assign({}, source);
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you have a very large codebase with extensive use of `new Object()`, migrating all instances may be time-consuming.
However, the migration is straightforward and the benefits of using object literals are significant.

## Further Reading

- [MDN: Object initializer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer)
- [MDN: Object() constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/Object)
- [MDN: Object.create()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="objectCalls" />
