---
description: "Use explicit octal syntax with the '0o' prefix instead of legacy octal numeric literals."
title: "octalNumbers"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="octalNumbers" />

Legacy octal numeric literals (e.g., `077`, `0123`) are a deprecated feature in JavaScript that can lead to confusion and errors.
They are forbidden in strict mode and are less readable than their modern alternatives.
The explicit octal syntax `0o` (e.g., `0o77`) introduced in ES6 is clearer and works in both strict and non-strict modes.
The digit `0` by itself is allowed as it represents zero, not an octal literal.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const permissions = 0755;
const value = 077;
const numbers = [01, 02, 03, 04, 05, 06, 07];
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const permissions = 0o755;
const value = 0o77;
const numbers = [0o1, 0o2, 0o3, 0o4, 0o5, 0o6, 0o7];

// The number zero is allowed
const zero = 0;

// Other number formats are allowed
const decimal = 123;
const hex = 0xff;
const binary = 0b1010;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you need to maintain compatibility with very old JavaScript engines that don't support the explicit octal syntax (ES6+), you might choose to disable this rule.
However, this is extremely rare in modern development, as ES6 has been widely supported since 2015.

## Further Reading

- [ECMAScript Specification: Numeric Literals](https://tc39.es/ecma262/#sec-literals-numeric-literals)
- [MDN: Numeric literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#numeric_literals)
- [MDN: Strict Mode in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="octalNumbers" />
