---
description: "Use explicit `undefined` values instead of array holes (sparse arrays)."
title: "sparseArrays"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="sparseArrays" />

Sparse arrays contain holes (empty slots) represented by commas without values between them.
These holes behave differently from `undefined` values and array methods treat them inconsistently, which can lead to unexpected behavior and bugs.
Using explicit `undefined` values makes the intent clear and ensures consistent behavior.

This rule reports on array literals with holes.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const items = [1, , 3]; // Hole at index 1
const values = [, 2, 3]; // Hole at index 0
const data = [1, , , 4]; // Multiple holes
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const items = [1, undefined, 3];
const values = [undefined, 2, 3];
const data = [1, undefined, undefined, 4];

// Trailing commas are allowed (not sparse)
const list = [1, 2, 3];

// Empty arrays are allowed
const empty = [];
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you intentionally use sparse arrays and understand their behavior, you might choose to disable this rule.
However, explicit `undefined` values are generally clearer and more predictable.

## Further Reading

- [MDN: Sparse Arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Indexed_collections#sparse_arrays)
- [MDN: Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="sparseArrays" />
