---
description: "Avoid using control flow statements in finally blocks."
title: "unsafeFinallyStatements"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="unsafeFinallyStatements" />

Control flow statements such as `return`, `throw`, `break`, and `continue` in `finally` blocks can cause unexpected behavior.
When these statements appear in a `finally` block, they override any control flow statements in the corresponding `try` or `catch` blocks.
The `finally` block always executes, and its control flow statements take precedence, which can lead to bugs that are difficult to diagnose.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
function processData() {
	try {
		return fetchData();
	} finally {
		return null;
	}
}

function handleError() {
	try {
		throw new Error("Original error");
	} catch (error) {
		throw error;
	} finally {
		throw new Error("Override error");
	}
}

for (let i = 0; i < items.length; i++) {
	try {
		processItem(items[i]);
	} finally {
		break;
	}
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
function processData() {
	try {
		return fetchData();
	} finally {
		cleanup();
	}
}

function handleError() {
	try {
		throw new Error("Original error");
	} catch (error) {
		console.error(error);
		throw error;
	} finally {
		cleanup();
	}
}

for (let i = 0; i < items.length; i++) {
	try {
		processItem(items[i]);
		break;
	} finally {
		cleanup();
	}
}
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

This rule enforces best practices that prevent subtle bugs related to control flow in exception handling.
There are very few legitimate use cases for control flow statements in `finally` blocks, so disabling this rule is not recommended.

## Further Reading

- [MDN: try...catch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch)
- [MDN: Control flow and error handling](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Control_flow_and_error_handling)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="unsafeFinallyStatements" />
