---
description: "Reports negating the left operand of `in` or `instanceof` relations."
title: "unsafeNegations"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="unsafeNegations" />

Negating the left-hand side of `in` or `instanceof` changes the meaning due to operator precedence.
Use `!(left in right)` or `!(left instanceof Right)` to negate the relation as intended.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
if ((!key) in object) {
}

if ((!object) instanceof Constructor) {
}

if ((!key) in object) {
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
if (!(key in object)) {
}

if (!(object instanceof Constructor)) {
}

if (key in object) {
}
if (object instanceof Constructor) {
}
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your codebase never negates `in` or `instanceof` relations, this rule may provide limited value.
However, keeping it enabled prevents subtle precedence mistakes that change control-flow conditions.

## Further Reading

- [MDN: `in` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in)
- [MDN: `instanceof` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof)
- [MDN: Operator precedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_precedence)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="unsafeNegations" />
